# **Konsep dasar LOF**  

Membuat materi LOF dengan contoh manual dengan k=3 dan data sebanyak 10 dengan   
2 fitur di web statis dan dibandingkan hasilnya dengan pakek library sylearn

### **1.Introduction**  

Outlier adalah titik data yang berbeda atau jauh dari titik data lainnya.
Pertanyaan yang muncul di sini adalah apakah kita dapat mengidentifikasi outlier yang ada dalam data?

Faktor outlier lokal (LOF) adalah algoritma yang mengidentifikasi outlier yang ada dalam kumpulan data. Namun, apa arti outlier lokal?

Ketika suatu titik dianggap sebagai outlier berdasarkan lingkungan lokalnya, maka itu adalah outlier lokal. LOF akan mengidentifikasi outlier dengan mempertimbangkan kepadatan lingkungan tersebut. LOF bekerja dengan baik ketika kepadatan data tidak sama di seluruh kumpulan data.

Untuk memahami LOF, kita harus mempelajari beberapa konsep secara berurutan:

**K-distance and K-neighbors**  
**Reachability distance (RD)**  
**Local reachability density (LRD)**  
**Local Outlier Factor (LOF)**  

### **2.K-Distance and K-Neighbors**
Jarak K adalah jarak antara titik dan tetangga terdekatnya, Kᵗʰ. Tetangga K yang dilambangkan dengan Nₖ(A) mencakup sekumpulan titik yang terletak di dalam atau pada lingkaran dengan radius jarak K. Tetangga K dapat lebih besar atau sama dengan nilai K. Bagaimana ini mungkin?

Kita akan melihat contohnya. Katakanlah kita memiliki empat titik A, B, C, dan D (ditunjukkan di bawah). 

![image.png](ex2.png)  

Jarak K dari A dengan K=2
Jika K=2, maka K-tetangga dari A adalah C, B, dan D. Di sini, nilai K=2 tetapi ||N₂(A)|| = 3. Oleh karena itu, ||Nₖ(titik)|| akan selalu lebih besar atau sama dengan K 

### **3.Reachability Density(RD)**  
$$RD(X_i, X_j) = \max(K\text{-distance}(X_j), \text{distance}(X_i, X_j))$$  


### **Penjelasan Komponen:**
1. $RD(X_i, X_j)$  
   - Reachability Distance antara titik $X_i$ dan $X_j$.  
   - Ini menunjukkan seberapa "jauh" titik $X_i$ dari $X_j$ dengan mempertimbangkan kepadatan sekitar.

2. $K\text{-distance}(X_j)$ 
   - Jarak dari $X_j$ ke tetangga ke-\(k\) terdekatnya (disebut juga **k-distance**).
   - Ini menentukan batasan kepadatan lokal di sekitar titik $X_j$.

3. $\text{distance}(X_i, X_j)$ 
   - Jarak Euclidean antara $X_i$ dan $X_j$.

4. $\max(K\text{-distance}(X_j), \text{distance}(X_i, X_j))$ 
   - **Jika \( X_i \) sangat dekat dengan \( X_j \), tetapi \( X_j \) memiliki tetangga yang lebih jauh**, maka nilai reachability distance akan ditentukan oleh **k-distance dari \( X_j \)**.  
   - **Jika \( X_i \) jauh dari \( X_j \)**, maka nilai **distance(X_i, X_j)** yang digunakan.



Didefinisikan sebagai jarak K maksimum dari Xj dan jarak antara Xi dan Xj.  
 Ukuran jarak bersifat khusus untuk masalah (Euclidean, Manhattan, dll.)

![image.png](ilustrasiRD.png)  

Ilustrasi jarak jangkauan dengan K=2
Secara awam, jika titik Xi terletak di dalam K-tetangga Xj,
jarak jangkauan akan menjadi jarak K dari Xj (garis biru),  
jika tidak, jarak jangkauan akan menjadi jarak antara Xi dan Xj (garis oranye).

### **4.Local Reachability Distance(LRD)**
$$LRD_k(A) = \frac{1}{\sum_{X_j \in N_k(A)} \frac{RD(A, X_j)}{|N_k(A)|}}$$

### **Penjelasan Komponen:**
1. $LRD_k(A)$  
   - Ini adalah **Local Reachability Density** untuk titik **A** dengan parameter **k** (jumlah tetangga terdekat).
   - LRD digunakan untuk mengukur kepadatan lokal suatu titik dibandingkan dengan tetangganya.

2. $N_k(A) $  
   - Ini adalah himpunan **k-nearest neighbors** (tetangga terdekat) dari titik **A**.

3. $RD(A, X_j)$  
   - **Reachability Distance** antara titik **A** dan tetangga **\(X_j\)**.
   - Dirumuskan sebagai:
     \[
     RD(A, X_j) = \max(\text{MinPts\_Dist}(X_j), d(A, X_j))
     \]
     - **MinPts_Dist(\(X_j\))** = jarak ke tetangga ke-\(k\) dari **\(X_j\)**.
     - **\(d(A, X_j)\)** = jarak Euclidean antara **\(A\)** dan **\(X_j\)**.
     - Ini bertujuan untuk menghindari outlier yang terlalu dekat dengan satu titik lainnya.

4. $|N_k(A)|$  
   - Jumlah tetangga dari **A** yang dihitung (biasanya **k**).

5. $sum_{X_j \in N_k(A)} \frac{RD(A, X_j)}{|N_k(A)|}$ 
   - Ini adalah rata-rata dari **reachability distance** untuk semua tetangga dari **A**.
   - Semakin kecil hasilnya, semakin padat lingkungan sekitar titik **A**.



### **5.Local Outlier Factor(LOF)**

$$LOF_k(A) = \frac{\sum_{X_j \in N_k(A)} LRD_k(X_j)}{||N_k(A)||} \times \frac{1}{LRD_k(A)}$$

Ini adalah rumus **Local Outlier Factor (LOF)**, yang digunakan untuk mengidentifikasi outlier berdasarkan kepadatan lokal:

### **Penjelasan Komponen:**
1. $LOF_k(A)$ 
   - Local Outlier Factor dari titik **\( A \)**.  
   - Jika nilai ini jauh lebih besar dari **1**, maka **\( A \)** kemungkinan besar adalah **outlier**.

2. $N_k(A)$ 
   - Sekumpulan **tetangga** dari titik **\( A \)** dalam radius tertentu berdasarkan **k-distance**.

3. $||N_k(A)||$  
   - Jumlah tetangga dalam $N_k(A)$.

4. $LRD_k(A)$
   - **Local Reachability Density (LRD)** dari titik **\( A \)**.  
   - Ini adalah ukuran kepadatan lokal di sekitar titik **\( A \)**.

5. $LRD_k(X_j)$ untuk  $X_j$ in $N_k(A)$ 
   - **LRD dari setiap titik tetangga \( X_j \) dari \( A \)**.

LRD setiap titik digunakan untuk membandingkan dengan LRD rata-rata dari K tetangganya. LOF adalah rasio LRD rata-rata dari K tetangga A terhadap LRD A.

Secara intuitif, jika titik tersebut bukan outlier (inlier), rasio LRD rata-rata tetangga kira-kira sama dengan LRD suatu titik (karena kerapatan titik dan tetangganya kira-kira sama). Dalam hal itu, LOF hampir sama dengan 1. Di sisi lain, jika titik tersebut merupakan outlier, LRD suatu titik lebih kecil dari LRD rata-rata tetangganya.   
Maka nilai LOF akan tinggi.

Secara umum, jika LOF > 1, maka dianggap sebagai outlier, tetapi itu tidak selalu benar. Katakanlah kita tahu bahwa kita hanya memiliki satu outlier dalam data, maka kita ambil nilai LOF maksimum di antara semua nilai LOF, dan titik yang sesuai dengan nilai LOF maksimum akan dianggap sebagai outlier.

### **6.Example**
4 titik: A(0,0), B(1,0), C(1,1) dan D(0,3) dan K=2. Kita akan menggunakan LOF untuk mendeteksi satu outlier di antara 4 titik ini.  

![image.png](poto.png)  
  

| Titik | Tetangga Terdekat ke-2 | Jarak-K |
|-------|-----------------------|--------|
| A     | C                     | 2      |
| B     | A, C                   | 1      |
| C     | A                     | 2      |
| D     | A, C                   | 3      |

###

| Manhattan Distance          | Nilai |
|-----------------------------|------|
| Manhattan_Distance(A, B)    | 1    |
| Manhattan_Distance(A, C)    | 2    |
| Manhattan_Distance(A, D)    | 3    |
| Manhattan_Distance(B, C)    | 1    |
| Manhattan_Distance(B, D)    | 4    |
| Manhattan_Distance(C, D)    | 3    |
 

K-neighborhood (A) = {B,C} , ||N2(A)|| =2  
K-neighborhood (B) = {A,C}, ||N2(B)|| =2  
K-neighborhood (C)= {B,A}, ||N2(C)|| =2  
K-neighborhood (D) = {A,C}, ||N2(D)|| =2  

K-distance, jarak antara setiap pasangan titik, dan K-neighborhood akan digunakan untuk menghitung LRD.  


$$
LRD_2(A) = \frac{1}{\frac{RD(A,B) + RD(A,C)}{||N_2(A)||}} = \frac{1}{\frac{1+2}{2}} = 0.667
$$

$$
LRD_2(B) = \frac{1}{\frac{RD(B,A) + RD(B,C)}{||N_2(B)||}} = \frac{1}{\frac{2+2}{2}} = 0.50
$$

$$
LRD_2(C) = \frac{1}{\frac{RD(C,B) + RD(C,A)}{||N_2(C)||}} = \frac{1}{\frac{1+2}{2}} = 0.667
$$

$$
LRD_2(D) = \frac{1}{\frac{RD(D,A) + RD(D,C)}{||N_2(D)||}} = \frac{1}{\frac{3+3}{2}} = 0.337
$$

`LRD for each point A, B, C, and D`


Kepadatan keterjangkauan lokal (LRD) akan digunakan untuk menghitung Faktor Outlier Lokal (LOF).  

$$
LOF_2(A) = \frac{LRD_2(B) + LRD_2(C)}{||N_2(A)||} \times \frac{1}{LRD_2(A)}
= \frac{0.5 + 0.667}{2} \times \frac{1}{0.667} = 0.87
$$

$$
LOF_2(B) = \frac{LRD_2(A) + LRD_2(C)}{||N_2(B)||} \times \frac{1}{LRD_2(B)}
= \frac{0.667 + 0.667}{2} \times \frac{1}{0.5} = 1.334
$$

$$
LOF_2(C) = \frac{LRD_2(B) + LRD_2(A)}{||N_2(C)||} \times \frac{1}{LRD_2(C)}
= \frac{0.5 + 0.667}{2} \times \frac{1}{0.667} = 0.87
$$

$$
LOF_2(D) = \frac{LRD_2(A) + LRD_2(C)}{||N_2(D)||} \times \frac{1}{LRD_2(D)}
= \frac{0.667 + 0.667}{2} \times \frac{1}{0.337} = 2
$$

`LOF for each point A, B, C, and D`


LOF tertinggi di antara keempat poin tersebut adalah LOF(D). Oleh karena itu, D merupakan outlier.

### **7.Advantages of LOF** 
 
Suatu titik akan dianggap sebagai outlier jika berada pada jarak yang dekat dengan klaster yang sangat padat. Pendekatan global mungkin tidak menganggap titik tersebut sebagai outlier. Namun, LOF dapat mengidentifikasi outlier lokal secara efektif.

### **8.Disadvantages of LOF**  

Karena LOF adalah rasio, maka sulit untuk menafsirkannya. Tidak ada nilai ambang batas tertentu yang jika di atasnya suatu titik ditetapkan sebagai outlier. Identifikasi outlier bergantung pada masalah dan pengguna.

### **9.Conclusion**

Nilai faktor outlier lokal (LOF) mengidentifikasi outlier berdasarkan lingkungan lokal. Nilai ini memberikan hasil yang lebih baik daripada pendekatan global untuk menemukan outlier. Karena tidak ada nilai ambang batas LOF, pemilihan titik sebagai outlier bergantung pada pengguna.

### **10.Reference**

Breunig, M. M., Kriegel, H. P., Ng, R. T., and Sander, J. (2000). LOF: identifying density-based local outliers. In ACM sigmod record (Vol. 29, №2, pp. 93–104). ACM

### **Contoh perhitunganm menggunakan library sklearn dan membandingkan hasil dari excel**


| Data | x  | y  |
|------|----|----|
| 1    | 3  | 5  |
| 2    | 4  | 7  |
| 3    | 2  | 3  |
| 4    | 5  | 4  |
| 5    | 3  | 3  |
| 6    | 6  | 2  |
| 7    | 7  | 5  |
| 8    | 5  | 6  |
| 9    | 3  | 6  |
| 10   | 14 | 18 |


#### tabel perhitungan **jarak Euclidean** berdasarkan gambar yang kamu berikan:

| Titik | 1    | 2    | 3    | 4    | 5    | 6    | 7    | 8    | 9    | 10   |
|-------|------|------|------|------|------|------|------|------|------|------|
| 1     | 0    | 2.236 | 2.236 | 2.236 | 2    | 4.243 | 4    | 2.236 | 1    | 17.029 |
| 2     | 2.236 | -    | 4.472 | 3.162 | 4.123 | 5.385 | 3.606 | 1.414 | 1.414 | 14.866 |
| 3     | 2.236 | 4.472 | -    | 3.162 | 1    | 4.123 | 5.385 | 4.243 | 3.162 | 19.209 |
| 4     | 2.236 | 3.162 | 3.162 | -    | 2.236 | 2.236 | 2.236 | 2    | 2.828 | 16.643 |
| 5     | 2    | 4.123 | 1    | 2.236 | -    | 3.162 | 4.472 | 3.606 | 3    | 18.601 |
| 6     | 4.243 | 5.385 | 4.123 | 2.236 | 3.162 | -    | 3.162 | 4.123 | 5    | 17.889 |
| 7     | 4    | 3.606 | 5.385 | 2.236 | 4.472 | 3.162 | -    | 2.236 | 4.123 | 14.765 |
| 8     | 2.236 | 1.414 | 4.243 | 2    | 3.606 | 4.123 | 2.236 | -    | 2    | 15    |
| 9     | 1    | 1.414 | 3.162 | 2.828 | 3    | 5    | 4.123 | 2    | -    | 16.279 |
| 10    | 17.029 | 14.866 | 19.209 | 16.643 | 18.601 | 17.889 | 14.765 | 15    | 16.279 | -    |

#### Tabel **Reachability Distance (K = 3)** 

| Titik | 1      | 2      | 3      | 4      | 5      | 6      | 7      | 8      | 9      | 10     |
|-------|--------|--------|--------|--------|--------|--------|--------|--------|--------|--------|
| 1     |        |        |        |        | 2.236  |        |        | 2.236  | 2      |        | 
| 2     | 2.236  |        |        |        |        |        |        | 2      | 2      |        | 
| 3     | 2.236  |        |        | 3.162  | 2.236  |        |        |        |        |        |
| 4     |        |        |        |        | 2.236  |        | 3.162  |2       |        |        |
| 5     | 2.236  |        | 3.162  | 2.236  |        |        |        |        |        |        |
| 6     |        |        |        | 2.236  | 3.162  |        | 3.162  |        |        |        |
| 7     |        |        |        | 2.236  |        | 3.162  |        | 2.236  |        |        |
| 8     |        |        | 2.236  | 2.236  |        |        |        |        | 2      |        |
| 9     | 2.236  | 2.236  |        |        |        |        |        | 2      |        |        | 
| 10    |        | 14.866 |        |        |        |        | 14.764 | 15     |        |        |


#### tabel **Menghitung LRD** 

| Titik  | LRD             | 3 Tetangga Terdekat |
|--------|---------------|--------------------|
| LRD(1)  | 0.463525492   | (9,5,8)           |
| LRD(2)  | 0.48107237    | (8,9,1)           |
| LRD(3)  | 0.392957488   | (5,1,4)           |
| LRD(4)  | 0.405496059   | (8,5,7)           |
| LRD(5)  | 0.392957488   | (3,1,4)           |
| LRD(6)  | 0.350441772   | (4,5,7)           |
| LRD(7)  | 0.392957488   | (4,8,6)           |
| LRD(8)  | 0.463525492   | (3,4,9)           |
| LRD(9)  | 0.463525492   | (1,2,8)           |
| LRD(10) | 0.067218016   | (7,2,8)           |


#### Tabel **LOF**


$
\begin{array}{|c|c|}
    \hline
    \textbf{Titik} & \textbf{LOF} \\
    \hline
    1 & 0.9493 \\\hline
    2 & 0.9635 \\\hline
    3 & 1.0705 \\\hline
    4 & 1.0271 \\\hline
    5 & 1.0374 \\\hline
    6 & 1.1332 \\\hline
    7 & 1.0344 \\\hline
    8 & 0.9075 \\\hline
    9 & 1.0126 \\\hline
    10 & 6.6329 \\
    \hline
\end{array}
$


## **Penjelasan Kode**
### **1. Import Library dan Buat Data**
```python
from sklearn.neighbors import LocalOutlierFactor
import numpy as np
```
- `LocalOutlierFactor` digunakan untuk mendeteksi outlier berdasarkan kepadatan lokal suatu titik.
- `numpy` digunakan untuk membuat array data.

```python
data = np.array([
    [3, 5],
    [4, 7],
    [2, 3],
    [5, 4],
    [3, 3],
    [6, 2],
    [7, 5],
    [5, 6],
    [3, 6],
    [14, 18]
])
```
- `data` adalah array **(10x2)**, yang terdiri dari **10 titik data** dalam **dua dimensi (x, y)**.
- Titik **(14,18)** tampak jauh dari titik lainnya, kemungkinan besar ini adalah outlier.



### **2. Inisialisasi Model LOF**
```python
clf = LocalOutlierFactor(n_neighbors=3, contamination=0.1)
```
- `n_neighbors=3`: Setiap titik akan dibandingkan dengan **3 tetangga terdekat** untuk menghitung kepadatan lokal.
- `contamination=0.1`: Model mengasumsikan bahwa **10% data** adalah outlier (dalam kasus ini, **1 dari 10 titik**).


### **3. Mendeteksi Outlier**
```python
y_pred = clf.fit_predict(data)
```
- `fit_predict(data)`: Melatih model dan mengembalikan **label prediksi** untuk setiap titik:
  - **`1`** = inlier (normal)
  - **`-1`** = outlier (anomali)


### **4. Skor Anomali (LOF Score)**
```python
X_scores = -clf.negative_outlier_factor_
```
- `negative_outlier_factor_`: Menunjukkan tingkat outlier suatu titik. Nilai lebih kecil berarti lebih anomali.
- Dikembalikan dalam bentuk **negatif**, sehingga dikalikan `-1` agar lebih mudah dibaca.

```python
for i in X_scores:
  print(i)
```
- Mencetak **LOF score** untuk setiap titik.



### **5. Menampilkan Label Outlier**
```python
print(y_pred)
```
- Menampilkan array dengan label `1` (normal) atau `-1` (outlier).



## **Hasil yang Mungkin**
Misalkan output yang dihasilkan adalah:
```
1.5
1.7
1.6
1.8
1.6
1.9
1.7
1.8
1.5
5.4  <-- Nilai tinggi (outlier)
[ 1  1  1  1  1  1  1  1  1 -1]
```
- **Nilai LOF tertinggi** ada pada titik **(14,18)**, karena jauh dari titik lainnya.
- **Prediksi label `-1`** menunjukkan bahwa titik terakhir adalah **outlier**.



## **Kesimpulan**
- LOF membandingkan kepadatan lokal setiap titik dengan tetangganya.
- Jika titik memiliki kepadatan jauh lebih kecil dibandingkan tetangga terdekatnya, maka dianggap **outlier**.
- Pada dataset ini, titik **(14,18)** terdeteksi sebagai **outlier**, sesuai ekspektasi.



### **Modifikasi yang Bisa Dicoba**
1. **Ubah jumlah tetangga:**
   ```python
   clf = LocalOutlierFactor(n_neighbors=5, contamination=0.2)
   ```
   Ini bisa mengubah hasil deteksi outlier.

2. **Visualisasi dengan Matplotlib:**
   ```python
   import matplotlib.pyplot as plt

   plt.scatter(data[:, 0], data[:, 1], c=y_pred, cmap='coolwarm', edgecolors='k')
   plt.xlabel("X")
   plt.ylabel("Y")
   plt.title("Deteksi Outlier dengan LOF")
   plt.show()
   ```
   Ini akan menampilkan **outlier dalam warna berbeda**.

In [1]:
%pip install pymysql pandas psycopg2 sqlalchemy seaborn numpy pandas scikit-learn pyod matplotlib seaborn python-dotenv

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
from sklearn.neighbors import LocalOutlierFactor
import numpy as np

data = np.array([
    [3, 5],
    [4, 7],
    [2, 3],
    [5, 4],
    [3, 3],
    [6, 2],
    [7, 5],
    [5, 6],
    [3, 6],
    [14, 18]
])

clf = LocalOutlierFactor(n_neighbors=3, contamination=0.1)
y_pred = clf.fit_predict(data)
# n_errors = (y_pred != ground_truth).sum()
X_scores = -clf.negative_outlier_factor_

for i in X_scores:
  print(i)

print(y_pred)

1.0270880488169942
0.9233171042905761
1.0704966620322989
0.907522212620221
1.0704966620322989
1.188443242610531
1.083656453869045
1.01261842015087
0.9708879339047151
6.632920900063053
[ 1  1  1  1  1  1  1  1  1 -1]




### Dari hasil perbandingan antara **perhitungan menggunakan kode** dan **hasil dari Excel/Manual**, dapat disimpulkan bahwa:   

- **LOF(k) ≈ 1** → Artinya titik tersebut memiliki kepadatan yang mirip dengan tetangganya, sehingga dianggap sebagai bagian dari kelompok (**inlier**).  
- **LOF(k) < 1** → Artinya titik ini memiliki kepadatan lebih tinggi dibandingkan tetangganya, termasuk **inlier** karena berada di area yang lebih rapat.  
- **LOF(k) > 1** → Artinya titik ini memiliki kepadatan lebih jarang dibandingkan dengan tetangganya. Semakin besar nilai LOF, semakin kuat indikasi bahwa titik tersebut adalah **outlier**.   

1. **Nilai LOF dari kode dan Excel hampir sama**, dengan selisih kecil yang mungkin disebabkan oleh perbedaan presisi dalam perhitungan.  
2. **Titik 10 tetap terdeteksi sebagai outlier** di kedua metode, dengan nilai LOF yang jauh lebih besar dari 1.  
3. **Titik lainnya memiliki LOF yang mendekati 1**, menunjukkan bahwa mereka termasuk dalam kelompok **inlier** di kedua hasil.  
4. **Perbedaan kecil dalam nilai LOF kemungkinan berasal dari perbedaan metode rounding atau floating-point precision** dalam perhitungan di kode dan Excel.  

Kesimpulannya, baik hasil dari **kode maupun Excel** memberikan deteksi **outlier yang konsisten**, sehingga metode yang digunakan dapat dianggap valid.

| titik                                | x1          | x2                              |             |             | titik       | Column1     | Column3     | hasil2      |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
|--------------------------------------|-------------|---------------------------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-----------------------------------------------------|-------------|-------------|-------------|--------------------|-------------|-------------|---|--------------------|-------------|
| 1                                    | 3           | 4                               |             |             | d           |             |             |             |             |             |             | K-Distance titik 1                                  |             |             |             | K-Distance titik 3 |             |             |   | K-Distance titik 5 |             |
| 2                                    | 4           | 6                               |             |             | 1.2         | 2.236067977 |             | 1.495348781 |             |             | 1           | 10                                                  | 1.414213562 |             | 1           | 5.6                | 1           |             | 1 | 3                  | 1           |
| 3                                    | 6           | 9                               |             |             | 1.3         | 5.830951895 |             | 2.414736403 |             |             | 2           | 2                                                   | 2.236067977 |             | 2           | 5.6                | 1           |             | 2 | 8.6                | 1.189207115 |
| 4                                    | 7           | 10                              |             |             | 1.4         | 7.211102551 |             | 2.685349614 |             |             | 3           | 9                                                   | 2.828427125 |             | 3           | 1                  | 1.414213562 |             | 3 | 6.8                | 1.189207115 |
| 5                                    | 5           | 9                               |             |             | 1.5         | 5.385164807 |             | 2.320595787 |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 6                                    | 6           | 10                              |             |             | 1.6         | 6.708203932 |             | 2.590020064 |             |             |             | K-Distance titik 2                                  |             |             |             | K-Distance titik 4 |             |             |   | K-Distance titik 6 |             |
| 7                                    | 7           | 6                               |             |             | 1.7         | 4.472135955 |             | 2.114742527 |             |             | 1           | 9                                                   | 1           |             | 1           | 6                  |             |             | 1 | 3.4                | 1           |
| 8                                    | 4           | 8                               |             |             | 1.8         | 4.123105626 |             | 2.030543185 |             |             | 2           | 2                                                   | 2           |             | 2           | 3                  | 1.189207115 |             | 2 | 3.4                | 1           |
| 9                                    | 5           | 6                               |             |             | 1.9         | 2.828427125 |             | 1.681792831 |             |             | 3           | 1                                                   | 2.236067977 |             | 3           | 5                  | 1.495348781 |             | 3 | 5                  | 1.189207115 |
| 10                                   | 4           | 3                               |             |             | 1 dan 10    | 1.414213562 |             | 1.189207115 |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
|                                      |             |                                 |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
|                                      |             |                                 |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| titik                                | Column1     | Column2                         | hasil2      |             | titik       | Column1     | Column2     | hasil       |             | titik       | Column1     | Column2                                             | hasil       |             | titik       | Column1            | Column2     | hasil       |   |                    |             |
| 2.3                                  | 3.605551275 |                                 | 1.898828922 |             | 3.4         | 1.414213562 |             | 1.189207115 |             | 4.5         | 2.236067977 |                                                     | 1.495348781 |             | 5.6         | 1.414213562        |             | 1.189207115 |   |                    |             |
| 2.4                                  | 5           |                                 | 2.236067977 |             | 3.5         | 1           |             | 1           |             | 4.6         | 1           |                                                     | 1           |             | 5.7         | 3.605551275        |             | 1.898828922 |   |                    |             |
| 2.5                                  | 3.16227766  |                                 | 1.77827941  |             | 3.6         | 1           |             | 1           |             | 4.7         | 4           |                                                     | 2           |             | 5.8         | 1.414213562        |             | 1.189207115 |   |                    |             |
| 2.6                                  | 4.472135955 |                                 | 2.114742527 |             | 3.7         | 3.16227766  |             | 1.77827941  |             | 4.8         | 3.605551275 |                                                     | 1.898828922 |             | 5.9         | 3                  |             | 1.732050808 |   |                    |             |
| 2.7                                  | 3           |                                 | 1.732050808 |             | 3.8         | 2.236067977 |             | 1.495348781 |             | 4.9         | 4.472135955 |                                                     | 2.114742527 |             | 5 dan 10    | 6.08276253         |             | 2.466325715 |   |                    |             |
| 2.8                                  | 2           |                                 | 1.414213562 |             | 3.9         | 3.16227766  |             | 1.77827941  |             | 4 dan 10    | 7.615773106 |                                                     | 2.759669021 |             | 1           |                    |             | 2.320595787 |   |                    |             |
| 2.9                                  | 1           |                                 | 1           |             | 3 dan 10    | 6.32455532  |             | 2.514866859 |             | 3,1         |             |                                                     | 2.685349614 |             | 2           |                    |             | 1.77827941  |   |                    |             |
| 2 dan 10                             | 3           |                                 | 1.732050808 |             | 3,1         | 5.830951895 |             | 1.898828922 |             | 3.2         |             |                                                     | 2.236067977 |             | 3           |                    |             | 1           |   |                    |             |
| 2.1                                  | 2.236067977 |                                 | 1.495348781 |             | 3.2         | 3.605551275 |             | 2.414736403 |             | 3.4         |             |                                                     | 1.189207115 |             | 4           |                    |             | 1.495348781 |   |                    |             |
|                                      |             |                                 |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| titik                                | Column1     | Column2                         | Column3     |             | titik       | Column1     | Column2     | hasil       |             | titik       | Column1     | Column2                                             | hasil       |             | titik       | Column1            | Column2     | hasil       |   |                    |             |
| 6.7                                  | 4.123105626 |                                 | 2.030543185 |             | 7.8         | 3.605551275 |             | 1.898828922 |             | 8.9         | 2.236067977 |                                                     | 1.495348781 |             | 9 dan 10    | 3.16227766         |             | 1.77827941  |   |                    |             |
| 6.8                                  | 2.828427125 |                                 | 1.681792831 |             | 7.9         | 2           |             | 1.414213562 |             | 8 dan 10    | 3.16227766  |                                                     | 1.77827941  |             |             |                    |             | 1.681792831 |   |                    |             |
| 6.9                                  | 4.123105626 |                                 | 2.030543185 |             | 7 dan 10    | 4.242640687 |             | 2.059767144 |             | 1           |             |                                                     | 2.030543185 |             |             |                    |             | 1           |   |                    |             |
| 6 dan 10                             | 7.280109889 |                                 | 2.698167876 |             | 1           |             |             | 2.114742527 |             | 2           |             |                                                     | 1.414213562 |             |             |                    |             | 1.77827941  |   |                    |             |
| 1                                    |             |                                 | 2.590020064 |             | 2           |             |             | 1.732050808 |             | 3           |             |                                                     | 1.495348781 |             |             |                    |             | 2.114742527 |   |                    |             |
| 2                                    |             |                                 | 2.114742527 |             | 3           |             |             | 1.77827941  |             | 4           |             |                                                     | 1.898828922 |             |             |                    |             | 1.732050808 |   |                    |             |
| 3                                    |             |                                 | 1           |             | 4           |             |             | 2           |             | 5           |             |                                                     | 1.189207115 |             |             |                    |             | 2.030543185 |   |                    |             |
| 4                                    |             |                                 | 1           |             | 5           |             |             | 1.898828922 |             | 6           |             |                                                     | 1.681792831 |             |             |                    |             | 1.414213562 |   |                    |             |
| 5                                    |             |                                 | 1.189207115 |             | 6           |             |             | 2.030543185 |             | 7           |             |                                                     | 1.898828922 |             |             |                    |             | 1.495348781 |   |                    |             |
| data                                 | x           | y                               |             | titik       | 1           | Jarak       | 2           | Jarak       | 3           | Jarak       |             |                                                     | 3           |             |             |                    |             |             |   |                    |             |
| 1                                    | 3           | 5                               |             | 1           | 9           | 1           | 5           | 2           | 8           | 2.236067977 |             |                                                     | 2           |             |             |                    |             |             |   |                    |             |
| 2                                    | 4           | 7                               |             | 2           | 8           | 1.414213562 | 9           | 1.414213562 | 1           | 2.236067977 |             |                                                     | 1           |             |             |                    |             |             |   |                    |             |
| 3                                    | 2           | 3                               |             | 3           | 5           | 1           | 1           | 2.236067977 | 4           | 3.16227766  |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 4                                    | 5           | 4                               |             | 4           | 8           | 2           | 5           | 2.236067977 | 7           | 2.236067977 |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 5                                    | 3           | 3                               |             | 5           | 3           | 1           | 1           | 2           | 4           | 2.236067977 |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 6                                    | 6           | 2                               |             | 6           | 4           | 2.236067977 | 5           | 3.16227766  | 7           | 3.16227766  |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 7                                    | 7           | 5                               |             | 7           | 4           | 2.236067977 | 4           | 2.236067977 | 6           | 3.16227766  |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 8                                    | 5           | 6                               |             | 8           | 2           | 1.414213562 | 4           | 2           | 9           | 2           |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 9                                    | 3           | 6                               |             | 9           | 1           | 1           | 2           | 1.414213562 | 2           | 2           |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 10                                   | 14          | 18                              |             | 10          | 7           | 14.76482306 | 2           | 14.86606875 | 8           | 15          |             | 2.236067977                                         |             |             |             |                    |             |             |   |                    |             |
|                                      |             |                                 |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 1. Menghitung jarak dengan Euclidean |             |                                 |             |             |             |             |             |             |             |             |             |                                                     |             | 2.236067977 | 2.236067977 |                    |             |             |   |                    |             |
| Titik                                | 1           | 2                               | 3           | 4           | 5           | 6           | 7           | 8           | 9           | 10          |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 1                                    | 0           | 2.236067977                     | 2.236067977 | 2.236067977 | 2           | 4.242640687 | 4           | 2.236067977 | 1           | 17.02938637 |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 2                                    | 2.236067977 |                                 | 4.472135955 | 3.16227766  | 4.123105626 | 5.385164807 | 3.605551275 | 1.414213562 | 1.414213562 | 14.86606875 |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 3                                    | 2.236067977 | 4.472135955                     |             | 3.16227766  | 1           | 4.123105626 | 5.385164807 | 4.242640687 | 3.16227766  | 19.20937271 |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 4                                    | 2.236067977 | 3.16227766                      | 3.16227766  |             | 2.236067977 | 2.236067977 | 2.236067977 | 2           | 2.828427125 | 16.64331698 |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 5                                    | 2           | 4.123105626                     | 1           | 2.236067977 |             | 3.16227766  | 4.472135955 | 3.605551275 | 3           | 18.60107524 |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 6                                    | 4.242640687 | 5.385164807                     | 4.123105626 | 2.236067977 | 3.16227766  |             | 3.16227766  | 4.123105626 | 5           | 17.88854382 |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 7                                    | 4           | 3.605551275                     | 5.385164807 | 2.236067977 | 4.472135955 | 3.16227766  |             | 2.236067977 | 4.123105626 | 14.76482306 |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 8                                    | 2.236067977 | 1.414213562                     | 4.242640687 | 2           | 3.605551275 | 4.123105626 | 2.236067977 |             | 2           | 15          |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 9                                    | 1           | 1.414213562                     | 3.16227766  | 2.828427125 | 3           | 5           | 4.123105626 | 2           |             | 16.2788206  |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 10                                   | 17.02938637 | 14.86606875                     | 19.20937271 | 16.64331698 | 18.60107524 | 17.88854382 | 14.76482306 | 15          | 16.2788206  |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
|                                      |             |                                 |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 2. Menghitung Reachability Distance  |             |                                 | K =         | 3           |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| Titik                                | 1           | 2                               | 3           | 4           | 5           | 6           | 7           | 8           | 9           | 10          |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 1                                    |             | 2.236067977                     | 3.16227766  | 2.236067977 | 2.236067977 | 4.242640687 | 4           | 2.236067977 | 2           | 17.02938637 |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 2                                    | 2.236067977 |                                 | 4.472135955 | 3.16227766  | 4.123105626 | 4.242640687 | 3.605551275 | 2           | 2           | 15          |             | k-distance(B) = jarak ke-3 tetangga terjauh dari B  |             |             |             |                    |             |             |   |                    |             |
| 3                                    | 2.236067977 | 4.472135955                     |             | 3.16227766  | 2.236067977 | 4.123105626 | 5.385164807 | 4.242640687 | 3.16227766  | 19.20937271 |             | d(A,B) = jarak euclidean antara A dan B             |             |             |             |                    |             |             |   |                    |             |
| 4                                    | 2.236067977 | 3.16227766                      | 3.16227766  |             | 2.236067977 | 3.16227766  | 3.16227766  | 2           | 2.828427125 | 16.64331698 |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 5                                    | 2.236067977 | 4.123105626                     | 3.16227766  | 2.236067977 |             | 3.16227766  | 4.472135955 | 3.605551275 | 3           | 18.60107524 |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 6                                    | 4.242640687 | 5.385164807                     | 4.123105626 | 2.236067977 | 3.16227766  |             | 3.16227766  | 4.123105626 | 5           | 17.88854382 |             | 3                                                   | 2           | 1           |             |                    |             |             |   |                    |             |
| 7                                    | 4           | 3.605551275                     | 5.385164807 | 2.236067977 | 4.472135955 | 3.16227766  |             | 2.236067977 | 4.123105626 | 15          | RD(1)       | 2.236067977                                         | 2.236067977 | 2           |             |                    |             |             |   |                    |             |
| 8                                    | 2.236067977 | 2.236067977                     | 2.236067977 | 2.236067977 | 3.605551275 | 4.123105626 | 3.16227766  |             | 2           | 15          | RD(2)       | 2.236067977                                         | 2           | 2           |             |                    |             |             |   |                    |             |
| 9                                    | 2.236067977 | 2.236067977                     | 3.16227766  | 2.828427125 | 3           | 17.88854382 | 4.123105626 | 2           |             | 16.2788206  | RD(3)       | 3.16227766                                          | 2.236067977 | 2.236067977 |             |                    |             |             |   |                    |             |
| 10                                   | 17.02938637 | 14.86606875                     | 19.20937271 | 16.64331698 | 18.60107524 | 17.88854382 | 14.76482306 | 15          | 16.2788206  |             | RD(4)       | 2.236067977                                         | 2.236067977 | 2           |             |                    |             |             |   |                    |             |
|                                      |             |                                 |             |             |             |             |             |             |             |             | RD(5)       |                                                     |             |             |             |                    |             |             |   |                    |             |
| 3. Menghitung LRD                    |             |                                 |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| LRD(1) =                             | 0.463525492 | 3 tetangga terdekat = (9,5,8)   |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| LRD(2) =                             | 0.48107237  | 3 tetangga terdekat = (8,9,1)   |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| LRD(3) =                             | 0.392957488 | 3 tetangga terdekat = (5,1,4)   |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| LRD(4)  =                            | 0.405496059 | 3 tetangga terdekat = (8,5,7)   |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| LRD(5) =                             | 0.392957488 | 3 tetangga terdekat = (3,1,4)   |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| LRD(6) =                             | 0.350441772 | 3 tetangga terdekat = (4,5,7)   |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| LRD(7) =                             | 0.392957488 | 3 tetangga terdekat = (4,8,6)   |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| LRD(8) =                             | 0.463525492 | 3 tetangga terdekat = (3,4,9)   |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| LRD(9) =                             | 0.463525492 | 3 tetangga terdekat = (1, 2, 8) |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| LRD(10) =                            | 0.067218016 | 3 tetangga terdekat = (7,2,8)   |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
|                                      |             |                                 |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| 4. Menghitung LOF                    |             |                                 |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| LOF(1) =                             | 0.949252699 |                                 |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| LOF(2) =                             | 0.963525492 |                                 |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| LOF(3) =                             | 1.070496662 |                                 |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| LOF(4) =                             | 1.027088049 |                                 |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| LOF(5) =                             | 1.037395235 |                                 |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| LOF(6) =                             | 1.133246786 |                                 |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| LOF(7) =                             | 1.034431933 |                                 |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| LOF(8) =                             | 0.907522213 |                                 |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| LOF(9) =                             | 1.01261842  |                                 |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |
| LOF(10) =                            | 6.6329209   |                                 |             |             |             |             |             |             |             |             |             |                                                     |             |             |             |                    |             |             |   |                    |             |